showdown/battle_modifier.py

```diff
@@ -598,17 +598,17 @@ def mega(battle, split_msg):
 
 def transform(battle, split_msg):
     if is_opponent(battle, split_msg):
-        transformed_into_name = normalize_name(split_msg[3].split(':')[1])
+        #transformed_into_name = normalize_name(split_msg[3].split(':')[1]) # Not that useful for nicked pokemon because both name and base_name are the species name
 
         battle_copy = deepcopy(battle)
         battle.opponent.active.boosts = deepcopy(battle.user.active.boosts)
 
         battle_copy.user.from_json(battle_copy.request_json)
 
-        if battle_copy.user.active.name == transformed_into_name or battle_copy.user.active.name.startswith(transformed_into_name):
-            transformed_into = battle_copy.user.active
-        else:
-            transformed_into = find_pokemon_in_reserves(transformed_into_name, battle_copy.user.reserve)
+        #if battle_copy.user.active.name == transformed_into_name or battle_copy.user.active.name.startswith(transformed_into_name): # Again, will fail with nicked pokemon
+        transformed_into = battle_copy.user.active
+        #else: # And when it fails:
+        #    transformed_into = find_pokemon_in_reserves(transformed_into_name, battle_copy.user.reserve) # This function will return None crashing the bot
```

Comment:
Excellent catch with the nicknames.

A couple things:

1. Test(s) should be added to the TestTransform class in  `tests/test_battle_modifiers.py` that test the situation where a nicknamed pokemon is transformed into. Take a look at some of the other tests in that class to see how it can be done.

2. That failing test doesn't have a mistake. It is testing an edge-case where ditto transforms into a pokemon that is switched out on the same turn ditto is switched-in. In other words, it tests the situation where the pokemon ditto transforms into is not on the field at the end of the turn. The stats _should_ be 1,2,3,4,5 because those are the stats that the `request_json` says p1's charmander has.

It is very un-intuitive because of the way I wrote this - but basically:

The battle object passed into `transform()` keeps track of the state of the battle from an observer's point of view. It also is a snapshot in time from when the `-transform` message came in the log - so you are right to think that we should only ever use the bot's active pokemon as `transformed_into`. However, the `battle_copy`'s state once the line `battle_copy.user.from_json()` line is run is **after the entire turn has completed** - so the Pokemon being transformed into could already be in the reserves here. This if/else check is for if that situation occurs.


I believe the only required change is to modify line 601 to be:
```
transformed_into_name = battle.user.active.name
```